drivers/gpio/gpio_tca6424a.c

```diff
@@ -0,0 +1,610 @@
+
+	ret = update_output_regs(dev, reg_out);
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int tca6424a_port_set_bits_raw(const struct device *dev, 
+				      gpio_port_pins_t mask)
+{
+	return tca6424a_port_set_masked_raw(dev, mask, mask);
+}
+
+static int tca6424a_port_clear_bits_raw(const struct device *dev,
+					gpio_port_pins_t mask)
+{
+	return tca6424a_port_set_masked_raw(dev, mask, 0);
+}
+
+static int tca6424a_port_toggle_bits(const struct device *dev,
+				     gpio_port_pins_t mask)
+{
+	struct tca6424a_drv_data * const drv_data =
+		(struct tca6424a_drv_data * const)dev->data;
+	uint32_t reg_out;
+	int ret;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	reg_out = drv_data->pins_state.output;
+	reg_out ^= mask;
+
+	ret = update_output_regs(dev, reg_out);
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int tca6424a_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	struct tca6424a_drv_data *drv_data = dev->data;
+	struct tca6424a_irq_state *irq = &drv_data->irq_state;
+
+	/* Device does not support level-triggered interrupts. */
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		irq->falling &= ~BIT(pin);
+		irq->rising &= ~BIT(pin);
+	} else { /* GPIO_INT_MODE_EDGE */
+		if (trig == GPIO_INT_TRIG_BOTH) {
+			irq->falling |= BIT(pin);
+			irq->rising |= BIT(pin);
+		} else if (trig == GPIO_INT_TRIG_LOW) {
+			irq->falling |= BIT(pin);
+			irq->rising &= ~BIT(pin);
+		} else if (trig == GPIO_INT_TRIG_HIGH) {
+			irq->falling &= ~BIT(pin);
+			irq->rising |= BIT(pin);
+		}
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	return 0;
+
+}
+
+static int tca6424a_manage_callback(const struct device *dev,
+				    struct gpio_callback *callback,
+				    bool set)
+{
+	struct tca6424a_drv_data *drv_data = 
+		(struct tca6424a_drv_data * const) dev->data;
+
+	return gpio_manage_callback(&drv_data->callbacks, callback, set);
+}
+
+static const struct gpio_driver_api tca6424a_drv_api = {
+	.pin_configure = tca6424a_pin_config,
+	.port_get_raw = tca6424a_port_get_raw,
+	.port_set_masked_raw = tca6424a_port_set_masked_raw,
+	.port_set_bits_raw = tca6424a_port_set_bits_raw,
+	.port_clear_bits_raw = tca6424a_port_clear_bits_raw,
+	.port_toggle_bits = tca6424a_port_toggle_bits,
+	.pin_interrupt_configure = tca6424a_pin_interrupt_configure,
+	.manage_callback = tca6424a_manage_callback,
+};
+
+/**
+ * @brief Initialization function of TCA6424A
+ *
+ * This sets initial input/output configuration and output states.
+ * The interrupt is configured if it is enabled.
+ *
+ * @param dev Device struct
+ *
+ * @retval 0 if successful.
+ * @retval Negative value for error code.
+ */
+static int tca6424a_init(const struct device *dev)
```

Comment:
In general in this function, why swallow errno values (e.g. gpio_pin_configure_dt()'s return value on line 503 is swallowed and replaced with -EIO) instead of propagating negative errno values higher up into the stack?

Reply:
The errno values now properly propagate up everywhere in this function.